clients/ethereum/erc20ContractsHolder_test.go

```diff
@@ -0,0 +1,168 @@
+		assert.Equal(t, errNilStatusHandler, err)
+	})
+	t.Run("should work", func(t *testing.T) {
+		args := createMockArgsContractsHolder()
+
+		ch, err := NewErc20SafeContractsHolder(args)
+		assert.Nil(t, err)
+		assert.False(t, check.IfNil(ch))
+	})
+}
+
+func TestBalanceOf(t *testing.T) {
+	t.Parallel()
+
+	t.Run("address does not exists on map nor blockchain", func(t *testing.T) {
+		expectedError := errors.New("no contract code at given address")
+		args := createMockArgsContractsHolder()
+		args.ethClient = &bridgeV2.ContractBackendStub{
+			CallContractCalled: func(ctx context.Context, call ethereum.CallMsg, blockNumber *big.Int) ([]byte, error) {
+				return nil, expectedError
+			},
+		}
+		ch, err := NewErc20SafeContractsHolder(args)
+		assert.Nil(t, err)
+		assert.False(t, check.IfNil(ch))
+
+		result, err := ch.BalanceOf(context.Background(), testsCommon.CreateRandomEthereumAddress(), testsCommon.CreateRandomEthereumAddress())
+		assert.Equal(t, expectedError, err)
+		assert.Nil(t, result)
+	})
+	t.Run("address exists only on blockchain", func(t *testing.T) {
+		var returnedBalance int64 = 1000
+		args := createMockArgsContractsHolder()
+		args.ethClient = &bridgeV2.ContractBackendStub{
+			CallContractCalled: func(ctx context.Context, call ethereum.CallMsg, blockNumber *big.Int) ([]byte, error) {
+				return convertBigToAbiCompatible(big.NewInt(returnedBalance)), nil
+			},
+		}
+		ch, err := NewErc20SafeContractsHolder(args)
+		assert.Nil(t, err)
+		assert.False(t, check.IfNil(ch))
+		assert.Equal(t, 0, len(ch.contracts))
+
+		result, err := ch.BalanceOf(context.Background(), testsCommon.CreateRandomEthereumAddress(), testsCommon.CreateRandomEthereumAddress())
+		assert.Nil(t, err)
+		assert.Equal(t, big.NewInt(returnedBalance), result)
+		assert.Equal(t, 1, len(ch.contracts))
+	})
+	t.Run("address exists also in contracts map", func(t *testing.T) {
+		var returnedBalance int64 = 1000
+		args := createMockArgsContractsHolder()
+		args.ethClient = &bridgeV2.ContractBackendStub{
+			CallContractCalled: func(ctx context.Context, call ethereum.CallMsg, blockNumber *big.Int) ([]byte, error) {
+				return convertBigToAbiCompatible(big.NewInt(returnedBalance)), nil
+			},
+		}
+		ch, err := NewErc20SafeContractsHolder(args)
+		contractAddress := testsCommon.CreateRandomEthereumAddress()
+		address1 := testsCommon.CreateRandomEthereumAddress()
+		address2 := testsCommon.CreateRandomEthereumAddress()
+		assert.Nil(t, err)
+		assert.False(t, check.IfNil(ch))
+		assert.Equal(t, 0, len(ch.contracts))
+
+		result, err := ch.BalanceOf(context.Background(), contractAddress, address1)
+		// first time the contract does not exists in the map, so it should add it
+		assert.Nil(t, err)
+		assert.Equal(t, big.NewInt(returnedBalance), result)
+		assert.Equal(t, 1, len(ch.contracts))
+
+		result, err = ch.BalanceOf(context.Background(), contractAddress, address2)
+		// second time the contract already exists in the map, so it should just use it
```

Comment:
this test doesn't quite test that the contract already exists in the map. Maybe add a helper function in an `export_test` file that directly get an object from the map?

Reply:
as we decided, a size check of 0 for the contracts map will be enough.